File: format/src/main/java/org/runestar/cache/format/Compressor.java
package org.runestar.cache.format;

import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream;
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream;
import org.runestar.cache.format.util.ByteBufferInputStream;
import org.runestar.cache.format.util.ByteBufferOutputStream;
import org.runestar.cache.format.util.IO;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.SequenceInputStream;
import java.io.UncheckedIOException;
import java.nio.ByteBuffer;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public enum Compressor {

    NONE {

        @Override public ByteBuffer decompress(ByteBuffer compressed) {
            return IO.getBuffer(compressed);
        }

        @Override public void compress(ByteBuffer buf, ByteBufferOutputStream dst) {
            dst.write(buf);
        }
    },

    BZIP2 {

        private static final int BLOCK_SIZE = 1;

        private final byte[] HEADER = {'B', 'Z', 'h', '0' + BLOCK_SIZE};

        @Override public ByteBuffer decompress(ByteBuffer compressed) {
            var out = new byte[compressed.getInt()];
            try (var in = new BZip2CompressorInputStream(new SequenceInputStream(new ByteArrayInputStream(HEADER), new ByteBufferInputStream(compressed)))) {
                IO.readBytes(in, out);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            return ByteBuffer.wrap(out);
        }

        @Override public void compress(ByteBuffer buf, ByteBufferOutputStream dst) {
            int start = dst.buf.position();
            int len = buf.remaining();
            try (var out = new BZip2CompressorOutputStream(dst, BLOCK_SIZE)) {
                IO.transferTo(buf, out);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            dst.buf.putInt(start, len);
        }
    },

    GZIP {

        @Override public ByteBuffer decompress(ByteBuffer compressed) {
            var out = new byte[compressed.getInt()];
            try (var in = new GZIPInputStream(new ByteBufferInputStream(compressed))) {
                IO.readBytes(in, out);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            return ByteBuffer.wrap(out);
        }

        @Override public void compress(ByteBuffer buf, ByteBufferOutputStream dst) {
            dst.writeInt(buf.remaining());
            try (var out = new GZIPOutputStream(dst)) {
                IO.transferTo(buf, out);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
    };

    abstract public ByteBuffer decompress(ByteBuffer compressed);

    abstract public void compress(ByteBuffer buf, ByteBufferOutputStream dst);

    public final ByteBuffer compress(ByteBuffer buf) {
        var out = new ByteBufferOutputStream();
        compress(buf, out);
        return out.buf.flip();
    }

    public byte id() {
        return (byte) ordinal();
    }

    public int headerSize() {
        return this == NONE ? 0 : Integer.BYTES;
    }

    public static Compressor of(byte id) {
        switch (id) {
            case 0: return NONE;
            case 1: return BZIP2;
            case 2: return GZIP;
        }
        throw new IllegalArgumentException("" + id);
    }

    public static Compressor best(ByteBuffer buf) {
        var out = new ByteBufferOutputStream();
        GZIP.compress(buf.duplicate(), out);
        int gzip = out.buf.position();
        out.buf.clear();
        BZIP2.compress(buf.duplicate(), out);
        int bzip2 = out.buf.position();
        int none = buf.remaining();
        if (none <= gzip && none <= bzip2) return NONE;
        if (gzip <= bzip2) return GZIP;
        return BZIP2;
    }
}

File: format/src/main/java/org/runestar/cache/format/Compressor.kt
package org.runestar.cache.format

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream
import java.io.SequenceInputStream

enum class Compressor(val headerLength: Int) {

    NONE(0) {

        override fun compress(buffer: ByteBuf): ByteBuf {
            return buffer.retainedDuplicate()
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            return buffer.retainedDuplicate()
        }
    },

    BZIP2(Integer.BYTES) {

        private val BLOCK_SIZE = 1

        private val HEADER = "BZh$BLOCK_SIZE".toByteArray(Charsets.US_ASCII)

        override fun compress(buffer: ByteBuf): ByteBuf {
            buffer.markReaderIndex()
            val decompressedSize = buffer.readableBytes()
            val outputBuffer = Unpooled.buffer()
            buffer.inputStream().use { input ->
                BZip2CompressorOutputStream(outputBuffer.outputStream(), BLOCK_SIZE).use { output ->
                    input.copyTo(output)
                }
            }
            buffer.resetReaderIndex()
            val header = outputBuffer.toArray(0, HEADER.size)
            check(header.contentEquals(HEADER)) {
                "invalid header; expected ${HEADER.toString(Charsets.US_ASCII)} but got ${header.toString(Charsets.US_ASCII)}"
            }
            outputBuffer.setInt(0, decompressedSize) // replace bzip2 header with decompressedSize, both 4 bytes
            return outputBuffer
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            buffer.markReaderIndex()
            val expectedDecompressedSize = buffer.readInt()
            val outputBuffer = Unpooled.buffer(expectedDecompressedSize)
            BZip2CompressorInputStream(SequenceInputStream(HEADER.inputStream(), buffer.inputStream())).use { input ->
               outputBuffer.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            buffer.resetReaderIndex()
            val decompressedSize = outputBuffer.readableBytes()
            check(decompressedSize == expectedDecompressedSize) {
                "invalid decompressed size; expected $expectedDecompressedSize but got $decompressedSize"
            }
            return outputBuffer
        }
    },

    GZIP(Integer.BYTES) {

        override fun compress(buffer: ByteBuf): ByteBuf {
            buffer.markReaderIndex()
            val outputBuffer = Unpooled.buffer()
            outputBuffer.writeInt(buffer.readableBytes())
            buffer.inputStream().use { input ->
                GzipCompressorOutputStream(outputBuffer.outputStream()).use { output ->
                    input.copyTo(output)
                }
            }
            buffer.resetReaderIndex()
            return outputBuffer
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            buffer.markReaderIndex()
            val expectedDecompressedSize = buffer.readInt()
            val outputBuffer = Unpooled.buffer(expectedDecompressedSize)
            GzipCompressorInputStream(buffer.inputStream()).use { input ->
                outputBuffer.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            buffer.resetReaderIndex()
            val decompressedSize = outputBuffer.readableBytes()
            check(decompressedSize == expectedDecompressedSize) {
                "invalid decompressed size; expected $expectedDecompressedSize but got $decompressedSize"
            }
            return outputBuffer
        }
    };

    abstract fun decompress(buffer: ByteBuf): ByteBuf

    abstract fun compress(buffer: ByteBuf): ByteBuf

    companion object {

        @JvmField val VALUES = values().asList()

        fun of(id: Byte): Compressor = VALUES[id.toInt()]
    }
}